pkg/catalog/walk_entry_iterator.go

```diff
@@ -0,0 +1,150 @@
+package catalog
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/treeverse/lakefs/pkg/ingest/store"
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type walkEntryIterator struct {
+	entries chan EntryWithMarker
+	walker  *store.WalkerWrapper
+
+	done   chan bool
+	closed *atomic.Bool
+	err    *atomic.Error
+
+	curr EntryWithMarker
+}
+
+// Mark stands for pagination information when listing objects from the blockstore.
+// It is used for server-client communication on the status of range ingestion.
+type Mark store.Mark
+
+type EntryWithMarker struct {
+	EntryRecord
+	Mark
+}
+
+// ErrItClosed is used to determine the reason for the end of the walk
+var ErrItClosed = errors.New("iterator closed")
+
+// buffer size of the buffer between reading entries from the blockstore Walk and passing it on
+const bufferSize = 100
+
+// WalkerFactory provides an abstraction for creating Walker
+type WalkerFactory interface {
+	GetWalker(ctx context.Context, opts store.WalkerOptions) (*store.WalkerWrapper, error)
+}
+
+func NewWalkEntryIterator(ctx context.Context, factory WalkerFactory, fromSourceURI, prepend, after, continuationToken string) (*walkEntryIterator, error) {
+	if prepend != "" && prepend[len(prepend)-1:] != "/" {
+		prepend += "/"
+	}
+
+	it := walkEntryIterator{
+		entries: make(chan EntryWithMarker, bufferSize),
+
+		done:   make(chan bool),
+		closed: atomic.NewBool(false),
+		err:    atomic.NewError(nil),
+	}
+	var err error
+	it.walker, err = factory.GetWalker(ctx, store.WalkerOptions{StorageURI: fromSourceURI})
+	if err != nil {
+		return nil, fmt.Errorf("creating object-store walker: %w", err)
+	}
+
+	go func() {
+		err := it.walker.Walk(ctx, store.WalkOptions{
+			After:             after,
+			ContinuationToken: continuationToken,
+		}, func(e store.ObjectStoreEntry) error {
+			if it.closed.Load() {
+				return ErrItClosed
+			}
+
+			it.entries <- EntryWithMarker{
+				EntryRecord: EntryRecord{
+					Path: Path(prepend + e.RelativeKey),
+					Entry: &Entry{
+						Address:      e.Address,
+						LastModified: timestamppb.New(e.Mtime),
+						Size:         e.Size,
+						ETag:         e.ETag,
+						AddressType:  Entry_FULL,
+						ContentType:  e.Address,
+					},
+				},
+				Mark: Mark(it.walker.Marker()),
+			}
+			return nil
+		})
+		if !errors.Is(err, ErrItClosed) {
+			it.err.Store(err)
+		}
+		close(it.entries)
+		close(it.done)
+	}()
+
+	return &it, nil
+}
+
+func (it *walkEntryIterator) Next() bool {
+	if it.err.Load() != nil || it.closed.Load() {
+		return false
+	}
+
+	var ok bool
+
+	select {
+	case it.curr, ok = <-it.entries:
+	case <-it.done:
+		// making sure not to miss entries
+		it.curr, ok = <-it.entries
+		if !ok {
+			// entries were exhausted
+			it.curr.Mark = Mark{
+				LastKey: it.curr.LastKey,
+				HasMore: false,
+			}
+		}
+	}
+
+	return ok
+}
+
+func (it *walkEntryIterator) SeekGE(Path) {
+	// unsupported
+}
```

Comment:
Think we will need a better way to handle this one - in other iterators, we set an error that this operation is not supported / panic.